File: ios/CoffeeBreak/Networking/APIClient.swift
import Foundation

struct ProgressDTO: Codable {
    let unlockedLevel: Int
    let xp: Int
    let coins: Int
    let lives: Int
    let livesUpdatedAt: Date
}

struct AuthResponse: Codable {
    let userId: String
    let token: String
    let progress: ProgressDTO?
}

enum APIError: Error {
    case invalidResponse
    case server(status: Int)
}

/// Thin wrapper around the backend REST API. See backend/README.md for the
/// endpoint list.
final class APIClient {
    static let shared = APIClient()

    private let baseURL = AppConfig.apiBaseURL
    private let session = URLSession.shared
    private var sessionToken: String? {
        get { UserDefaults.standard.string(forKey: "coffeebreak.sessionToken") }
        set { UserDefaults.standard.set(newValue, forKey: "coffeebreak.sessionToken") }
    }

    private init() {}

    /// Creates/loads the device-local account. Call this on first launch
    /// before any other authenticated request, if the player hasn't
    /// connected a real account.
    func authenticateWithDevice() async throws -> AuthResponse {
        let response: AuthResponse = try await post("/auth/device", body: ["deviceId": DeviceIdentity.current])
        sessionToken = response.token
        return response
    }

    /// Sign in with Apple. Pass `linkDeviceId` to attach the current
    /// device-local account's progress to the Apple ID instead of loading a
    /// separate Apple-linked account.
    func authenticateWithApple(identityToken: String, linkDeviceId: String? = nil) async throws -> AuthResponse {
        var body: [String: String] = ["identityToken": identityToken]
        if let linkDeviceId {
            body["linkDeviceId"] = linkDeviceId
        }
        let response: AuthResponse = try await post("/auth/apple", body: body)
        sessionToken = response.token
        return response
    }

    func fetchProgress() async throws -> ProgressDTO? {
        struct Wrapper: Codable { let progress: ProgressDTO? }
        let wrapper: Wrapper = try await get("/progress")
        return wrapper.progress
    }

    /// Reports a finished level so the server can apply the XP reward
    /// (04-systemes-progression-et-xp.md). Not called from any screen yet.
    func completeLevel(stars: Int, movesRemaining: Int) async throws -> (progress: ProgressDTO, xpGained: Int) {
        struct Body: Encodable { let stars: Int; let movesRemaining: Int }
        struct Wrapper: Decodable { let progress: ProgressDTO; let xpGained: Int }
        let wrapper: Wrapper = try await post("/progress/complete-level", body: Body(stars: stars, movesRemaining: movesRemaining))
        return (wrapper.progress, wrapper.xpGained)
    }

    /// Spends one life. Not called from any screen yet -- the trigger
    /// (level start vs. loss) isn't defined until later spec files.
    func consumeLife() async throws -> ProgressDTO {
        struct Wrapper: Decodable { let progress: ProgressDTO }
        let wrapper: Wrapper = try await post("/progress/consume-life", body: Empty())
        return wrapper.progress
    }

    // MARK: - Low-level helpers

    private struct Empty: Encodable {}

    private func get<Response: Decodable>(_ path: String) async throws -> Response {
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.httpMethod = "GET"
        return try await send(request)
    }

    private func post<Body: Encodable, Response: Decodable>(_ path: String, body: Body) async throws -> Response {
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)
        return try await send(request)
    }

    private func send<Response: Decodable>(_ request: URLRequest) async throws -> Response {
        var request = request
        if let sessionToken {
            request.setValue("Bearer \(sessionToken)", forHTTPHeaderField: "Authorization")
        }

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }
        guard (200..<300).contains(http.statusCode) else {
            throw APIError.server(status: http.statusCode)
        }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode(Response.self, from: data)
    }
}

File: mobile/lib/networking/api_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

import '../models/app_config.dart';
import '../models/device_identity.dart';

class ProgressDTO {
  final int unlockedLevel;
  final int xp;
  final int coins;
  final int lives;
  final DateTime livesUpdatedAt;

  ProgressDTO({
    required this.unlockedLevel,
    required this.xp,
    required this.coins,
    required this.lives,
    required this.livesUpdatedAt,
  });

  factory ProgressDTO.fromJson(Map<String, dynamic> json) => ProgressDTO(
        unlockedLevel: json['unlockedLevel'] as int,
        xp: json['xp'] as int,
        coins: json['coins'] as int,
        lives: json['lives'] as int,
        livesUpdatedAt: DateTime.parse(json['livesUpdatedAt'] as String),
      );
}

class AuthResponse {
  final String userId;
  final String token;
  final ProgressDTO? progress;

  AuthResponse({required this.userId, required this.token, this.progress});

  factory AuthResponse.fromJson(Map<String, dynamic> json) => AuthResponse(
        userId: json['userId'] as String,
        token: json['token'] as String,
        progress: json['progress'] == null
            ? null
            : ProgressDTO.fromJson(json['progress'] as Map<String, dynamic>),
      );
}

class ApiException implements Exception {
  final int status;
  ApiException(this.status);

  @override
  String toString() => 'ApiException(status: $status)';
}

/// Thin wrapper around the backend REST API. See backend/README.md for the
/// endpoint list.
class ApiClient {
  ApiClient._();
  static final ApiClient shared = ApiClient._();

  static const String _tokenKey = 'coffeebreak.sessionToken';
  final Uri _baseUrl = Uri.parse(AppConfig.apiBaseUrl);
  final http.Client _http = http.Client();

  Future<String?> get _sessionToken async =>
      (await SharedPreferences.getInstance()).getString(_tokenKey);

  Future<void> _setSessionToken(String token) async {
    await (await SharedPreferences.getInstance()).setString(_tokenKey, token);
  }

  /// Creates/loads the device-local account. Call this on first launch
  /// before any other authenticated request, if the player hasn't
  /// connected a real account.
  Future<AuthResponse> authenticateWithDevice() async {
    final deviceId = await DeviceIdentity.current();
    final response = AuthResponse.fromJson(
      await _post('/auth/device', {'deviceId': deviceId}),
    );
    await _setSessionToken(response.token);
    return response;
  }

  /// Sign in with Apple. Pass [linkDeviceId] to attach the current
  /// device-local account's progress to the Apple ID instead of loading a
  /// separate Apple-linked account.
  Future<AuthResponse> authenticateWithApple(
    String identityToken, {
    String? linkDeviceId,
  }) async {
    final body = <String, String>{'identityToken': identityToken};
    if (linkDeviceId != null) body['linkDeviceId'] = linkDeviceId;
    final response = AuthResponse.fromJson(await _post('/auth/apple', body));
    await _setSessionToken(response.token);
    return response;
  }

  /// Sign in with Google -- the Android-launch counterpart to Apple sign-in
  /// (backend/src/routes/auth.ts POST /auth/google). Pass [linkDeviceId] to
  /// attach the current device-local account's progress to the Google
  /// account instead of loading a separate Google-linked account.
  Future<AuthResponse> authenticateWithGoogle(
    String idToken, {
    String? linkDeviceId,
  }) async {
    final body = <String, String>{'idToken': idToken};
    if (linkDeviceId != null) body['linkDeviceId'] = linkDeviceId;
    final response = AuthResponse.fromJson(await _post('/auth/google', body));
    await _setSessionToken(response.token);
    return response;
  }

  Future<ProgressDTO?> fetchProgress() async {
    final wrapper = await _get('/progress');
    final progress = wrapper['progress'];
    return progress == null ? null : ProgressDTO.fromJson(progress as Map<String, dynamic>);
  }

  /// Reports a finished level so the server can apply the XP reward
  /// (04-systemes-progression-et-xp.md). Not called from any screen yet.
  Future<(ProgressDTO progress, int xpGained)> completeLevel({
    required int stars,
    required int movesRemaining,
  }) async {
    final wrapper = await _post('/progress/complete-level', {
      'stars': stars,
      'movesRemaining': movesRemaining,
    });
    return (
      ProgressDTO.fromJson(wrapper['progress'] as Map<String, dynamic>),
      wrapper['xpGained'] as int,
    );
  }

  /// Spends one life. Not called from any screen yet -- the trigger
  /// (level start vs. loss) isn't defined until later spec files.
  Future<ProgressDTO> consumeLife() async {
    final wrapper = await _post('/progress/consume-life', {});
    return ProgressDTO.fromJson(wrapper['progress'] as Map<String, dynamic>);
  }

  // MARK: - Low-level helpers

  Future<Map<String, dynamic>> _get(String path) => _send('GET', path);

  Future<Map<String, dynamic>> _post(String path, Map<String, dynamic> body) =>
      _send('POST', path, body: body);

  Future<Map<String, dynamic>> _send(
    String method,
    String path, {
    Map<String, dynamic>? body,
  }) async {
    final uri = _baseUrl.resolve(path);
    final token = await _sessionToken;
    final headers = <String, String>{
      'Content-Type': 'application/json',
      if (token != null) 'Authorization': 'Bearer $token',
    };

    final response = method == 'GET'
        ? await _http.get(uri, headers: headers)
        : await _http.post(uri, headers: headers, body: jsonEncode(body ?? {}));

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw ApiException(response.statusCode);
    }
    return jsonDecode(response.body) as Map<String, dynamic>;
  }
}
